### ISSUE TITLE
GRN early stopping does not mirror arboreto's `EarlyStopMonitor`

### ISSUE THREAD

**Issue by AtlasMaroc** [Ekin-Kahraman/rustscenic#95]
### Summary

difference in implementation between rustscenic and scenic+

### Minimal reproducible example

```python
Just indexing the repo using claude and comparing it to scenic+
```

### Current behaviour
crates/rustscenic-grn/src/gbm.rs documents itself as:

//! Early stopping mirrors arboreto's `EarlyStopMonitor` (window=25):
//! if in-bag `MSE[i] >= MSE[i-window]`, stop. Disabled with window=0.
and implements exactly that — a two-point comparison of in-bag mean squared error:

if window > 0 && scratch.mse_history.len() > window {
    let past = scratch.mse_history[scratch.mse_history.len() - 1 - window];
    if mse_inbag >= past { break; }
}
### Upstream behaviour
arboreto/core.py, EarlyStopMonitor.__call__:

if current_round >= self.window_length - 1:
    lo, hi = self.window_boundaries(current_round)
    return np.mean(regressor.oob_improvement_[lo: hi]) < 0

### Two differences, not one:

* Data split — arboreto monitors oob_improvement_, computed on the ~10% of rows held out by subsample=0.9. RustScenic monitors in-bag loss.
* Statistic — arboreto takes the mean improvement over the window. RustScenic compares two individual points 25 apart.

### Impact

In-bag training loss in gradient boosting decreases near-monotonically, so the RustScenic condition fires much later than arboreto's, or never. Three consequences:

* n_estimators stops behaving as a ceiling. arboreto's own comment on SGBM_KWARGS says n_estimators: 5000 # can be arbitrarily large precisely because the OOB monitor terminates it early. In RustScenic it is close to a literal tree count, and cost is roughly linear in it.
* Importances are rescaled. n_fit is the denormalization multiplier (gbm.rs, per arboreto/core.py:168), so a different stopping point changes the magnitude of every importance, not just which trees contribute.
* Plausibly a significant share of the reported GRN Spearman of 0.611.

### Suggested fix

The information needed is already in the loop. It walks every sample and already tracks in-bag membership via a sorted cursor into sample_idx; the OOB rows are the untaken else, and the pre-update residual is the "loss before" term:

let old_resid = scratch.residuals[k];
*p += cfg.learning_rate * predict_binned(&scratch.tree, binned, k);
scratch.residuals[k] = y[k] - *p;
if is_inbag { inbag_pos += 1; }
else {
    oob_before += old_resid * old_resid;
    oob_after  += scratch.residuals[k] * scratch.residuals[k];
    n_oob += 1;
}
Push (oob_before - oob_after) / n_oob into the history and stop when the trailing-window mean is < 0. Runtime cost is negligible.

Because this changes every published importance value, it should land as an option — e.g. early_stop_criterion="oob" | "inbag" — with the current behaviour reachable so committed artefacts stay reproducible.

Note one residual mismatch: subsample_rows_into uses Bernoulli sampling, so the OOB set size varies per iteration, where sklearn draws a fixed-size permutation subsample.

Validation available in-repo
validation/parity_v0310/ has committed arboreto output on PBMC3k. Re-run grn_parity_v0310.py before and after, and log the mean/median n_fit per target under both criteria.

### rustscenic version

main branch as of 13 August 

### Python version

3.12.5

### OS

Ubuntu 24.04 x86_64

### Full traceback or error output

```shell

```

### Checks

- [x] I searched existing issues for duplicates.
- [x] I tried upgrading to the latest rustscenic release.

**Comment by Ekin-Kahraman** [comment]
Thanks for identifying this. The fix is now in draft PR #97.

The root cause was that RustScenic used its historical two-point in-bag MSE rule rather than arboreto's strict trailing window over `oob_improvement_`. The PR makes `early_stop_mode="arboreto"` the default, uses fixed-size in-bag samples, retains the stopping tree, and keeps `legacy_inbag` as an explicit reproducibility mode. The reference container now pins scikit-learn 1.7.2 because modern and historical scikit-learn calculate OOB improvement differently.

On the exact committed PBMC3k parity fixture, the final corrected rerun produced 974,784 edges and 362,608 fitted trees, versus 363,178 in the arboreto reference. Median and p95 fitted-tree counts match at 26 and 30. The corrected parquet reproduced SHA-256 `87570e0956c9244ded26f995de4c5a82bd8246c86958bebaad8b3c4badb45ddd` after the final telemetry refactor. Fixed-universe importance Spearman increased from 0.6146 to 0.6363 and global top-10k Jaccard from 0.1964 to 0.2340; no scientific threshold was weakened.

The same stop distribution was also checked on the public #69 lymphoma source against scikit-learn 1.7.2 for six fixed targets. Different tree builders do not give exact per-target equality, but the median absolute fitted-tree difference was one.

Local Rust, Python and harness checks are green. Publication-grade Minerva validation remains pending because this workstation cannot resolve the Minerva host outside its network/VPN boundary. I am keeping this issue open while PR #97 is reviewed; it is configured to close this issue when merged.
